.github/CONTRIBUTING.md

```diff
@@ -69,102 +67,142 @@ There are few ways you can play with the build:
 
-There are 2 ways to run the test suite, either one-off or continuously:
+> If the test pages take any screenshots, run them in an automated manner to save the screenshots.
 
-> To run Web Driver test, locally, download [ChromeDriver](https://sites.google.com/a/chromium.org/chromedriver/downloads) and extract to project root. Then set environment variable `WEBCHAT_TEST_ENV=chrome-local` before running Jest.
+### Running tests automatically
 
--  For one-off testing, `node_modules/.bin/jest`
--  For continuous testing, `npm test`
-   -  To speed up continuous testing, you can install [watchman](https://facebook.github.io/watchman/)
+For test environment convergence and stability, Web Chat uses Docker for hosting the test environment. Please install the following components:
 
-If your development box has less than 4 cores, you will need to reduce the number of simultaneous test agents:
+-  On Windows, install [Windows Subsystem for Linux 2 (a.k.a. WSL 2)](https://aka.ms/wsl2)
+-  [Docker Desktop](https://www.docker.com/get-started)
+-  (Optional) Install [watchman](https://facebook.github.io/watchman/) to improve Jest performance
 
--  For one-off testing: run `node_modules/.bin/jest --maxWorkers 1`
--  For continuous testing: run `npm test -- --maxWorkers 1`
+Run `npm run docker`. It will start a Docker Compose with Selenium Grid and 4 nodes of Chrome.
 
-Our CI pipeline run tests with 4 agents simultaneously. If new tests are added, please make sure they can run simultaneously.
+Then, run `npm test` to start Jest. When Jest run the test pages, it will take screenshots and save it under [`/__tests__/__image_snapshots__/html`](https://github.com/microsoft/BotFramework-WebChat/tree/main/__tests__/__image_snapshots__/html).
 
-### Troubleshooting the test suite
+### Troubleshooting test suites
 
-We run test suite on every commit and requires 100% test pass. If the test suite did not complete successfully, they are likely:
+We run test suites on every commit and requires 100% test pass. If test suites did not complete successfully, they are likely:
 
+-  New changes causing fail in existing tests
 -  Intermittent services issue
 -  Test reliability issue, please see [#2938](https://github.com/microsoft/BotFramework-WebChat/issues/2938)
 -  Polluted development environment, for example:
    -  Outdated `node_modules` content
    -  Outdated Node.js or NPM
 
-When the test suite failed:
+When the test suites failed:
+
+-  Identify if it is related to new changes or not
+   -  Clone a clean repository and run the test suites without any changes
+-  Identify if it is intermittent issue or not
+   -  Run the test suites again
+
+If existing test suites failed without any changes:
+
+-  Test suites always fail
+   1. Service could be updated and causing a break in our test suites, please [file an issue](https://github.com/microsoft/BotFramework-WebChat/issues/new/choose)
```

Comment:
- Potential causes:
      - Service changes may be causing test suite breakage
      - Service reliability issues (e.g. DirectLine, MockBot, etc) may be causing intermittent failure
   - Please [file an issue](https://github.com/microsoft/BotFramework-WebChat/issues/new/choose) to report these failures